functorch/_src/make_functional.py

```diff
@@ -241,79 +192,175 @@ def fun(weights, buffers, data):
-                 param_names_map, buffer_names_map):
+    def __init__(
+        self,
+        stateless_model,
+        param_module_names,
+        param_modules,
+        param_names,
+        buffer_module_names,
+        buffer_modules,
+        buffer_names
+    ):
         super(FunctionalModuleWithBuffers, self).__init__()
         self.stateless_model = stateless_model
+        self.param_module_names = param_module_names
+        self.param_modules = param_modules
         self.param_names = param_names
+        self.buffer_module_names = buffer_module_names
+        self.buffer_modules = buffer_modules
         self.buffer_names = buffer_names
 
-        self.all_names_map = dict(param_names_map)
-        self.all_names_map.update(buffer_names_map)
-
     @staticmethod
-    def _create_from(model, disable_autograd_tracking=False):
+    def _create_from(model):
         # TODO: We don't need to copy the model to create a stateless copy
         model_copy = copy.deepcopy(model)
-        params, param_names, param_names_map = extract_weights(model_copy)
-        buffers, buffer_names, buffer_names_map = extract_buffers(model_copy)
+        param_container = list(extract_weights(model_copy))
+        if len(param_container):
+            param_module_names, param_modules, param_names, params = zip(*param_container)
+        else:
+            param_module_names, param_modules, param_names, params = tuple(), tuple(), tuple(), tuple()
         if disable_autograd_tracking:
             for param in params:
                 param.requires_grad_(False)
+
+        buffer_container = list(extract_buffers(model_copy))
+        if len(buffer_container):
+            buffer_module_names, buffer_modules, buffer_names, buffers = zip(*buffer_container)
+        else:
+            buffer_module_names, buffer_modules, buffer_names, buffers = tuple(), tuple(), tuple(), tuple()
         return (
-            FunctionalModuleWithBuffers(model_copy, param_names, buffer_names,
-                                        param_names_map, buffer_names_map),
+            FunctionalModuleWithBuffers(
+                model_copy,
+                param_module_names,
+                param_modules,
+                param_names,
+                buffer_module_names,
+                buffer_modules,
+                buffer_names),
             params,
             buffers,
         )
 
     def forward(self, params, buffers, *args, **kwargs):
-        # Temporarily load the state back onto self.stateless_model
-        old_state = _swap_state(
-            self.stateless_model,
-            self.all_names_map,
-            list(params) + list(buffers))
+        old_states = _swap_state(
+            self.param_modules + self.buffer_modules,
+            self.param_names + self.buffer_names,
+            list(params) + list(buffers)
+        )
+        old_params = old_states[:len(self.param_modules)]
+        old_buffers = old_states[len(self.param_modules):]
+
```

Comment:
Just to make sure I understand why this is faster: is it because we no longer need to traverse through the module to find the submodules; we've already made the submodules directly available to swap their parameters out?

Reply:
Exactly! 
Instead of going through a tree of param names, we just flatten it and go through a single list of modules, one-level names and values.